### ISSUE TITLE
Divide and conquer the Inventory class

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#463]
**Is your feature request related to a problem? Please describe.**
Simplify the Inventory class

**Describe the solution you'd like**
1. We may separate the use of Inventory to various class. Currently, it holds 
- Unit class such as Person, Vehicle, Equipment
- Amount Resources
- Item Resources

**Describe alternatives you've considered**
Leave it alone.

**Additional context**
See previous threads of issues and discussions
1. https://github.com/mars-sim/mars-sim/discussions/293#discussioncomment-1420250. 
2. https://github.com/mars-sim/mars-sim/issues/348. 
3. https://github.com/mars-sim/mars-sim/issues/306.

**Comment by mokun** [comment]
@bevans2000 

Now, I start working on removing the use of Inventory for EVASuit.

It could use a lightweight form of inventory, much like those container types (Bag, Specimen Box, etc.).

Currently, EVASuit tracks only 3 specific resources (oxygen, water and co2) and so it doesn't need to own a full blown Inventory class.

**Comment by bevans2000** [comment]
The current Inventory pays 3 roles

1. Holder of AmountResources - references by AmountResource
2. Holder of Equipment - referenced by EquipmentType
3. Holder of Parts - referenced by Part

Each of these could be represented as a dedicated Interface with separate implementations.  Then certain Unit sub classes would implement the appropriate interface, e.g Equipment would just be HolderAmountResource. A settlement would be all 3.; however the implementation for Settlement would also record demand automatically.

Each of the interfaces follow the same behaviour and would have just:
- How much is stored?
- Remove an amount - return how much was actually removed.
- Add an amount - return how much was actually added
- Transfer and amount to another instance of the same  interface.
- Drop the exception throwing if there is not enough stored; method return how much was actually transferred

**Comment by mokun** [comment]
> Then certain Unit sub classes would implement the appropriate interface, e.g Equipment would just be HolderAmountResource

Sounds good. Please do go ahead with this plan to simplify the object graph of Inventory

**Comment by mokun** [comment]
@bevans2000 ,

After this [commit](https://github.com/mars-sim/mars-sim/commit/5bdcfb56be5c796d4976c5c58913973a9c51cdc0), currently, there's a bug that creates the following weight issue on the EVA suit :

```
00-Adir-01:558.743 (Severe) Inventory : [New Pompeii - Pranav Jai] EVA Suit 063 - Had a mass of 1401.6 kg - too much to put on 'Pranav Jai' to carry. Remaining Gen Cap : 59.9 kg. (Gen Cap : 60.0 kg)
00-Adir-01:558.743 (Severe) Inventory : [Unknown] EVA Suit 063 - Could not be stored.
00-Adir-01:558.884 (Severe) Inventory : [New Pompeii - Aradhana Sandeep] EVA Suit 064 - Had a mass of 1331.4 kg - too much to put on 'Aradhana Sandeep' to carry. Remaining Gen Cap : 64.9 kg. (Gen Cap : 65.0 kg)
00-Adir-01:558.884 (Severe) Inventory : [Unknown] EVA Suit 064 - Could not be stored.
```

It is caused by the method canStoreUnit(Unit unit, boolean allowDirty) in Inventory. Somehow it reports that an EVA suit weighs too much and can't be put on. Everything looks okay to me and I have no idea why.

Can you take a look and figure out why ? Thanks !

**Comment by mokun** [comment]
Also, the getMass() in Unit looks normal as follows: 

```
	/**
	 * Gets the unit's mass including inventory mass.
	 * 
	 * @return mass of unit and inventory
	 * @throws Exception if error getting the mass.
	 */
	public double getMass() {
		double invMass = 0;
		if (this instanceof Equipment) {
			if (this instanceof EVASuit)
				invMass = ((EVASuit)this).getStoredMass();
			else
				invMass = ((Equipment)this).getStoredMass();
		}
		else {
			if (inventory == null) { 
				logger.severe(this + ": inventory is null.");
			}
			else 
				inventory.getTotalInventoryMass(false);
			if (invMass == 0)
				return baseMass;
		}
		
		return baseMass + invMass;
	}
```

Guess there's a better way to code getMass() to avoid all those instanceof. 

It would call getStoredMass() in EVASuit to obtain the weight of the just resources in an EVA suit.

The baseMass of an empty EVAsuit was computed at the start of the sim as the summation of all the parts the compose the suit, namely, the array EVASUIT_PARTS in ItemResourceUtil.

**Comment by bevans2000** [comment]
Sure. I think this code highlights the problem we face where there are a lot of instanceof to cater for Units that do not have an Inventory.

**Comment by mokun** [comment]
> Sure. I think this code highlights the problem we face where there are a lot of instanceof to cater for Units that do not have an Inventory.

Yea. Say, after we fix this problem for handling resources for EVA suit , we can for sure take a closer look at optimizing it and getting rid of those instanceof's.

**Comment by bevans2000** [comment]
OK so I looked at this. The problem is in setQuantity you use 'add' instead of 'set' so it keeps adding to the quantities instead of updating the existing value. The was also a mistake in the ```Unit.getMass``` method where it was not using the mass of the inventory. There are still a few warnings about a EVA Suit being too heavy to store but these cases look like the Person is just carrying too much stuff as the EVA Suit mass is reasonable. For what I can see these problem cases are when the Person has a small capacity; maybe it is something wrong with the change to do carrying capacity related to age?

Having the separate parallel lists is a dangerous approach because they can get out of synch. Why not just bit the bullet and use a Map? It will be efficient. Each Map Entry would be a simple POJO holding the current amount & capacity. 

What I can see is the work you've done in EVASuit & Equipment containers is very similar. One just holds a fixed set of resources and the Containers just hold a single resource. I think what could be done to improve this is to create a new class ```ResourceStorageImpl``` that encapsulates the logic which moves the resource logic out of EVASuit into a new place. The supported Resources would be passed in the constructor. Hence the same class can also be used for the Bag, Box, etc BUT only a single resource is used.
I would then put an interface ```ResourceStorage``` holding the key methods to retrieve & store and then implement this on the appropriate Units. Then many of the instanceof code can be replaced by checking against the interface instead which is always more acceptable.  Then the Units would use a delegate pattern to use the ```ResourceStorageImpl``` class. This class would be easy to Unit test in isolation as well which is a bonus.
The other improvement I would make to the simple class approach is to hold the total mass as a property instead of recalculating each time. The mass will be constant and only change when resources are added or removed. The size of an extra double is nothing compared to the processing avoided by not continuously adding up the volumes. Asking for current volumes is more frequent than adding/removing.
Then further ahead; this new class could be backfilled into the existing Inventory to replace logic there.

**Comment by mokun** [comment]
> Having the separate parallel lists is a dangerous approach because they can get out of synch. Why not just bit the bullet and use a Map? It will be efficient. Each Map Entry would be a simple POJO holding the current amount & capacity.

okay. I'll switch to Map

> create a new class ResourceStorageImpl that encapsulates the logic which moves the resource logic 
> I would then put an interface ResourceStorage holding the key methods to retrieve & store 

I'll let you lead the way to implement this correctly. Is it okay ?

I suppose this can be extend to hold Item Resources as well in future.

We can use this in replace of the Inventory inside the vehicle as vehicle allows a few other resources to be stored.

> The other improvement I would make to the simple class approach is to hold the total mass as a property instead of recalculating each time

Sure. It shouldn't need to be recalculated.

**Comment by bevans2000** [comment]
> > Having the separate parallel lists is a dangerous approach because they can get out of synch. Why not just bit the bullet and use a Map? It will be efficient. Each Map Entry would be a simple POJO holding the current amount & capacity.
> 
> okay. I'll switch to Map
> 
> > create a new class ResourceStorageImpl that encapsulates the logic which moves the resource logic
> > I would then put an interface ResourceStorage holding the key methods to retrieve & store
> 
> I'll let you lead the way to implement this correctly. Is it okay ?
> 
> I suppose this can be extend to hold Item Resources as well in future.
Well I would go simple and have a different class and interface for Item Resources. This would remove the need to always check the range if the resource id before using it. Same pattern as Resources.
> 
> We can use this in replace of the Inventory inside the vehicle as vehicle allows a few other resources to be stored.
Yes. But small steps are needed here since Inventory is so key at the moment.

I should have said I did a commit earlier that fixed the EVA suit problem.

**Comment by mokun** [comment]
Yes. Thanks for fixing that.

**Comment by bevans2000** [comment]
Looking at the stack trace the question is why are we updating the Inventory if we are just getting the capacity?

**Comment by bevans2000** [comment]
> Sure. I have seen something similar but rarely. It happens when the simulation adds Equipment. There is a flaw in the logic of events as well. The Unit constructor fire an event but of course the object is not fully constructed so a NPE appears in the monitor tool. That is why I logged #309 .

You can enable diagnostics via the diagnostics command in the Console. This will dump to a text file whenever a Unit is created.

**Comment by bevans2000** [comment]
Super. .My thoughts were the methods that use the explicit subclasses like Bag make it harder to understand. Certainly increases the coupling. 

The ```EquipmentFactory``` is another related bad example of using Class. So I may try and remove the dependencies of the Class references and replacement with EquiomentType. Then dropping in a generic Container will be easier.

**Comment by mokun** [comment]
@bevans2000 

I changed out most of the instanceof in the last [commit](https://github.com/mars-sim/mars-sim/commit/a1c0c5b142fa9173d5150293eda0e50e019f2e87), plus a few other minor things.

Somehow it can't pass the LoadVehicleTest in maven. 

Can you take a look to see what went wrong ?

**Comment by mokun** [comment]
> Somehow it can't pass the LoadVehicleTest in maven.

If I run this test by itself, it passes. Don't know why it can't pass when running together with other tests.

**Comment by mokun** [comment]
I also keep having these logs : 

```java
00-Adir-01:451.631 (Severe) Inventory [x232] : [Tian Cheng] grey water could not be totally retrieved. Remaining: 0.02295414699169962
00-Adir-01:459.964 (Severe) Inventory [x7989] : [Schiaparelli Point] grey water could not be totally retrieved. Remaining: 0.024622359810129044
```

They obviously came from retrieveAmountResource(int resource, double amount) in Inventory:  

```java
				if (remainingAmount > SMALL_AMOUNT_COMPARISON) {
					logger.log(getOwner(), Level.SEVERE, 30_000,
							ResourceUtil.findAmountResourceName(resource)
							+ " could not be totally retrieved. Remaining: " + remainingAmount);
				}
```

**Comment by bevans2000** [comment]
> @bevans2000
> 
> I changed out most of the instanceof in the last [commit](https://github.com/mars-sim/mars-sim/commit/a1c0c5b142fa9173d5150293eda0e50e019f2e87), plus a few other minor things.
> 
Super. 

> Somehow it can't pass the LoadVehicleTest in maven.
> 
> Can you take a look to see what went wrong ?
Sure. My guess is the tests are not isolated and one is affecting another.

**Comment by bevans2000** [comment]
> I also keep having these logs :
> 
> ```
> 00-Adir-01:451.631 (Severe) Inventory [x232] : [Tian Cheng] grey water could not be totally retrieved. Remaining: 0.02295414699169962
> 00-Adir-01:459.964 (Severe) Inventory [x7989] : [Schiaparelli Point] grey water could not be totally retrieved. Remaining: 0.024622359810129044
> ```
> 
> They obviously came from retrieveAmountResource(int resource, double amount) in Inventory:
> 
> ```java
> 				if (remainingAmount > SMALL_AMOUNT_COMPARISON) {
> 					logger.log(getOwner(), Level.SEVERE, 30_000,
> 							ResourceUtil.findAmountResourceName(resource)
> 							+ " could not be totally retrieved. Remaining: " + remainingAmount);
> 				}
> ```

I gave alse been seeing this message.

**Comment by mokun** [comment]
So, after fixing the rounding error, we still have the following:

```
00-Adir-01:362.771 (Severe) Inventory [x2417] : [Boring Town] black water could not be totally retrieved. Remaining: 8.074323062406902E-4
00-Adir-01:406.026 (Severe) Inventory [x2890] : [Boring Town] solid waste could not be totally retrieved. Remaining: 2.3357863144819966E-4
00-Adir-01:449.286 (Severe) Inventory [x2518] : [Boring Town] black water could not be totally retrieved. Remaining: 8.074323062406902E-4
00-Adir-01:492.835 (Severe) Inventory [x3475] : [Boring Town] grey water could not be totally retrieved. Remaining: 5.712625590074991E-5
00-Adir-01:536.780 (Severe) Inventory [x1180] : [Boring Town] grey water could not be totally retrieved. Remaining: 1.3250707262841497E-5
00-Adir-01:580.064 (Severe) Inventory [x3734] : [Boring Town] regolith could not be totally retrieved. Remaining: 0.004037161531203451
```

I'm beginning to suspect it may have to do with commenting out those cache methods in the last few commits.

**Comment by mokun** [comment]
btw, I'm converting those parallel lists in EVASuit to maps as the next incremental change.

This way, it will facilitate the creation of the future MicroInventory class.

**Comment by mokun** [comment]
> It only occurs with the ResourceProcess task.

Yea. it occurs a lot with grey water. Sometimes with black water and regolith. So it would have to do with calling the retrieveAmountResource from those resource processes that happens in every frame.

**Comment by mokun** [comment]
> a simple POJO containing stored & capacity as the value in the Map.

Yes. I'm creating a POJO MicroInventory class as we speak.

**Comment by bevans2000** [comment]
Super. You'll be able to unit test this in isolation. Potentially retro-fit it to Inventory as well are replace the heavyweight AmountResoruceStorage potentially.

**Comment by bevans2000** [comment]
Person and Robots can only hold Equipment  so is that something we can exploit.

**Comment by bevans2000** [comment]
> @bevans2000 ,
> 
> I just did the commit of incorporating MicroInventory for all equipment.
> 
> However, EVASuit has unresolved issues in storing amount resources.
> 
> I'm still wondering why loadEVASuit() in ExitAirlock can't store oxygen and water properly.
> 
> Can you figure out what went wrong ?

Sure I'll have a look

**Comment by mokun** [comment]
> But we need remove all overloaded methods in EVAsuit

But to do so, we'll need to reconcile the difference between how EVASuit recognize one of the 3 amount resources versus how Equipment select only one amount resource.

In future, I suppose we can make a Large Bag suitable for storing ONLY ONE type of amount resource, OR for storing a few different types of item resources.

When that happen, we'll need to modify those methods again in Equipment.

**Comment by bevans2000** [comment]
> btw, you mentioned making micro inventory loosely coupled by using interface and adding a concrete implementation of that interface, right ?
Yes this could be a later stages. So for example MicroInventory could implement ```ResourceHolder``` then any owner of a MicroInventory could implement the same interface and then use delegation to access the MicroInventory. This would allow resources/items to be stored and retrieved directly from the owner without having underlying class. Also our transfer patterns can be based on the responsiblilities of a class and not the class type.

**Comment by mokun** [comment]
> I feel this is wrong because you are holding 2 concepts in 1 class now. For example the `getResourcesStored` method was simple before as it returned Resources stored but now it it returning a mixture of resource ids & items ids. In some places this is going to cause a problem. Also we reference Resource & Items using the same integer sequence which means everywhere we have the same if else code looking at the range of the identifier. It would make sense to separate these because they are separate concepts.
> We will still need if else in places to check what is being stored/retrieved if using a single inventory.

but at some point along the way, it will need to discern if it's an amount resource or item resource.

So it's either before calling a person's micro inventory's method or after. 

It seems, iither way, an if-else is needed, unless we we come with a different way to automatically tell the resource types.

Back in the old Inventory class before 3.1.0, there is no resource id given to these resource types (amount resource,  item resource, settlement, vehicle, robot types, equipment types). 

In those days, there's a lot of issues with serialization, because, the object graph is very messy. If it works, it works like magic. If it breaks, it's hard to know why it works.

**Comment by bevans2000** [comment]
Super. What are you doing about Person having Equipment? The current contained Units of Inventory is over the top for Person (& Robot) since they can only carry Equipment. At the moment a Person could be assigned a Settlement !!!
Do we need a simple EquipmentHolder that Person and Robots use?

**Comment by mokun** [comment]
> What are you doing about Person having Equipment

I just did the commit. It's a long one. Hope it doesn't break anything big.

Currently, the change would account for a person having multiple equipment, each having its own micro inventory as well.   

I suppose the changes in Robot would a little simpler than those of Person.



> Do we need a simple EquipmentHolder that Person and Robots use?

How would EquipmentHolder work ?

**Comment by bevans2000** [comment]
> Currently, the change would account for a person having multiple equipment, each having its own micro inventory as well.
> 
> I suppose the changes in Robot would a little simpler than those of Person.
> 
> > Do we need a simple EquipmentHolder that Person and Robots use?
> 
> How would EquipmentHolder work ?
I don't know but sounds like a good 🤣 
I was more thinking about avoiding duplicate code in Robot.

**Comment by bevans2000** [comment]
Well something is not working right. When loading vehicles sometimes it warns there is not enough capacity to store it but there is plenty. However when debugging the Inventory attached to the Vehicle the ResourceStorage object reports there is 400kg of oxygen already loaded ???? No idea where this has come from. There could be 2 reasons:
1. It has been loaded by the controller but not being reported as there because of this Inventory caching
2. The oxygen is already there but it's not being reported. It seems the caching is not being updated with the content sof the ResourceStore.
Also when loading it is trying to load the extra Oxygen into containers that are loaded into the Vehicle but the vehcile should have enough capacity. I haven't a clue what is going on.

**Comment by mokun** [comment]
I added this interface [EquipmentOwner](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-core/src/main/java/org/mars_sim/msp/core/equipment/EquipmentOwner.java) to encapsulate all the new methods that have to do with managing resources and equipment in Person.

Next, we should be able to use EquipmentOwner on Robot, Vehicle, ConstructionSite and, at last, Settlement.

Any changes you'd like to make ? 

Also, you said earlier you would prefer to use one micro inventory for amount resources and one for item resources. I suppose we can do that as well.

**Comment by bevans2000** [comment]
Next step is to decide what to do with Robot ? I think
1.  Isolate the logic in Person that manages the Equipment, e.g. EquipmentHolder class
2. Move Robot to under Unit out of Equipment
3. Robot implements EquipentOwner and uses the existing MicroInvenrory and EquipmentHolder

Does that make sense?

**Comment by mokun** [comment]
> 1. Isolate the logic in Person that manages the Equipment, e.g. EquipmentHolder class

how is EquipmentHolder referred to in Person ? 

is it instantiated this way, just like Inventory ?

`EquipmentHolder equipmentHolder  = new EquipmentHolder() ?`

A lot of methods in Person are now there because they are required by EquipentOwner.

So, how's EquipentOwner related to EquipentHolder ?

**Comment by mokun** [comment]
> I'm going to fix my merge tonight and then commit it.

Sure. I'll wait till after you fix that and merge and commit first before I touch it again.

**Comment by mokun** [comment]
> May be the advantage with what we are currently working on is that we create one implementation of storeAmountResource() method for Person and a different implementation of the same method for Robot.

One real example is EVAsuit's getAmountResourceCapacity versus Equipment's getAmountResourceCapacity

From Equipment,

```
	/**
	 * Gets the capacity of a particular amount resource
	 * 
	 * @param resource
	 * @return capacity
	 */
	public double getAmountResourceCapacity(int resource) {
		if (isResourceSupported(resource)) {
			return getTotalCapacity();
		}
		return 0;
	}
```


From EVASuit,

```
	/**
	 * Gets the capacity of a particular amount resource
	 * 
	 * @param resource
	 * @return capacity
	 */
	@Override
	public double getAmountResourceCapacity(int resource) {
		// Note: this method is different from 
		// Equipment's getAmountResourceCapacity 
		return microInventory.getCapacity(resource);
	}

```

**Comment by bevans2000** [comment]
So I now believe that there is no need to have any of the Resource methods in EVASuit any more. It should be able to use the base method in Equipment. If you look at GenericContainer I have overridden a few methods to enforce that it only holds one resource.
Plus the new UnitTest checks this as well.
We need to reduce the methods that Person is exposing. We have multiple methods that are slightly similar and one not called.  I got into a mess so I've aborted my merge and will try it again.
Plus there is a problem in MicroInventory. The getResourcesStored method is return ids for bit AmountRessources & ItemResources so it could cause problems. Also do we need the capacity; I think remaining capacity is good enough everywhere.

I think we should leave Robot as is for now using Inventory and focus on getting Person working well. If we get Person working well.

**Comment by mokun** [comment]
btw, i forgot we still have a class called InventoryUtil that have some static methods that we can. 

So, may be sometimes it's good having static methods instead of methods that uses internal data members.

**Comment by bevans2000** [comment]
> Should we leave out giving each building a MicroInventory ?

Yes Building should store any resources in its parent Settlement.

**Comment by mokun** [comment]
> Yes Building should store any resources in its parent Settlement.

The thing is that in buildings.xml, each building may define a storage space capable of storing specific resources as well as a general stock-capacity for anything. 

Below is the storage element for *Lander Hab*: 

```xml
			<!-- This building is capable of storing the following resources. -->
				<!-- stock-capacity: the building's default capacity (kg) for 
				     all the resources. -->
			<storage stock-capacity="10000">

				<!-- The building's storage capacity for oxygen. -->
				<!-- resource: the specific resources that the building may 
				     store on top of the aformentioned stock-capacity -->
				<!-- capacity: the additional capacity (kg) for a resource. -->

				<resource-storage resource="toxic waste" capacity="500.0" />
				<resource-storage resource="solid waste" capacity="500.0" />
				<resource-storage resource="food waste" capacity="500.0" />
				<resource-storage resource="grey water" capacity="500.0" />
				<resource-storage resource="black water" capacity="500.0" />

				<resource-storage resource="food" capacity="500.0" />
				<resource-storage resource="water" capacity="500.0" />
				<resource-storage resource="oxygen" capacity="500.0" />
				<resource-storage resource="carbon dioxide" capacity="500.0" />
				<resource-storage resource="carbon monoxide" capacity="500.0" />
				<resource-storage resource="argon" capacity="500.0" />
				<resource-storage resource="nitrogen" capacity="500.0" />

				<resource-storage resource="ice" capacity="500.0" />
				<resource-storage resource="regolith" capacity="500.0" />

				<resource-storage resource="chlorine" capacity="100.0" />
				<resource-storage resource="sodium hypochlorite" capacity="100.0" />
				<resource-storage resource="acetic acid bacteria" capacity="100.0" />
				<resource-storage resource="toilet tissue" capacity="100.0" />
				<resource-storage resource="paper" capacity="100.0" />
				<resource-storage resource="napkin" capacity="100.0" />

				<!-- The initial resources in the building. -->
				<!-- resource: the resource initially stored in this building. -->
				<!-- amount: the amount (kg) of the resource. -->

				<!-- life support: 5000 kg  -->
				<resource-initial resource="water" amount="5000.0" />
				<resource-initial resource="oxygen" amount="5000.0" />
				<resource-initial resource="food" amount="1000.0" />

				<!-- other gases: 200 kg  -->
				<resource-initial resource="nitrogen" amount="130.0" />
				<resource-initial resource="argon" amount="50.0" />
				<resource-initial resource="chlorine" amount="20.0" />

				<!-- kitchen supplies: 600 kg -->				
				<resource-initial resource="toilet tissue" amount="50.0" />
				<resource-initial resource="paper" amount="50.0" />
				<resource-initial resource="napkin" amount="50.0" />
				<resource-initial resource="pulp" amount="50.0" />
				<resource-initial resource="honey" amount="50.0" />
				<resource-initial resource="yeast" amount="50.0" />
				<resource-initial resource="epsom salt" amount="50.0" />
				<resource-initial resource="table salt" amount="50.0" />
				<resource-initial resource="sugar" amount="50.0" />
				<resource-initial resource="mustard" amount="50.0" />
				<resource-initial resource="sesame seed" amount="50.0" />
				<resource-initial resource="sesame oil" amount="50.0" />
				<resource-initial resource="garlic oil" amount="50.0" />
				<resource-initial resource="peanut oil" amount="50.0" />
											
				<!-- chemicals: 210 kg  -->								
				<resource-initial resource="bleaching chemical" amount="50.0" />
				<resource-initial resource="baking powder" amount="50.0" />
				<resource-initial resource="bisphenol-a" amount="50.0" />
				<resource-initial resource="sodium hypochlorite" amount="50.0" />
				<resource-initial resource="immune booster" amount="50.0" />
				<resource-initial resource="lime" amount="10.0" />

				<!-- organisms: 40 kg -->	
				<resource-initial resource="acetic acid bacteria" amount="10.0" />
				<resource-initial resource="nitrosomonas spp" amount="10.0" />
				<resource-initial resource="nitrospira spp" amount="10.0" />
				<resource-initial resource="rhizobia" amount="10.0" />

				<!-- greenhouse supplies: 100 kg  -->
				<resource-initial resource="fertilizer" amount="50.0" />
				<resource-initial resource="soil" amount="50.0" />

			</storage>
```

**Comment by mokun** [comment]
For dirty amount resources, we should probably make that bag to remember what it has carried. We shouldn't allow that bag to be used for other resources.

**Comment by bevans2000** [comment]
> For dirty amount resources, we should probably make that bag to remember what it has carried. We shouldn't allow that bag to be used for other resources.
Yes that is how container work currently.  We can easy change it so different containers can be reused.

**Comment by mokun** [comment]
Anything you see troublesome with the Vehicle class using micro inventory ?

**Comment by mokun** [comment]
seems it's straight forward for Vehicle class to make the switch. It already has a Crewable interface class that both Rover and LightUtilityVehicle implements for handling which person or robot are on board.

**Comment by bevans2000** [comment]
> Anything you see troublesome with the Vehicle class using micro inventory ?

Just making sure Person is working correctly beforehand. Then create a new class to support the containment logic and then slot that directly into Vehicle. We have to avoid copy and pasting.

**Comment by mokun** [comment]
> seems it's straight forward for Vehicle class to make the switch. It already has a Crewable interface class that both Rover and LightUtilityVehicle implements for handling which person or robot are on board.

The `Airlock` abstract class may need some work.

**Comment by bevans2000** [comment]
Airlock has different concrete class for Building and Vehicle so differences can be pushed into those.

**Comment by mokun** [comment]
btw, I removed the getIndex() methods and the resourceIDs list in both Person and Vehicle. It's one step closer to having the methods as EVASuit and Equipment.

**Comment by mokun** [comment]
> If we put the EquipmentList to microinventory we are just reinventing the current Inventory class.

Yes. That's why I can't group these methods into a new class unless I drag EquipmentList with me. It's becoming more and more like the same old Inventory.

**Comment by bevans2000** [comment]
@mokun On my branch I've got all the UnitTest (plus some new ones) working. But I've come across the same problem s you with no one being able to get to the surface. Did you have any luck finding out why? I think the problem is in ```Unit.transfer```.
I think the best and cleanest answer is to move the transfer method to the specific Unit subclasses that can be transferred, e.g. Equipment, Person, Vehicle & Robot would have their own implementation. Make sense because then it is not possible to transfer a Settlement or Building; sounds correct. We could also use the new EquipmentOwner check on the origin & destination.
I plan to create a Pull Request tomorrow night with my changes after merging anything you've done on master.

**Comment by bevans2000** [comment]
Good.

**Comment by bevans2000** [comment]
> In terms of transfer, we have the following scenarios: S = Settlement M = Mars Surface V = Vehicle P = Person R = Robot E = Equipment
> 
> For Person
> 
> Origin	Destination
> S	M
> M	S
> S	V
> V	S
> V	M
> M	V
> For Equipment
> 
> Origin	Destination
> S	P
> P	S
> V	P
> P	V
> R	S
> S	R
> R	V
> V	R
> For Vehicle
> 
> Origin	Destination
> S	M
> M	S

So ,looking at these tables it feels like we could move the transfer to the specific Unit classes. Also we can consider Person and Vehicle the same if we use EquipmentOwner. I will commit my changes later today.

**Comment by mokun** [comment]
btw, one change is that an equipment can be transferred from a person to a person and from a robot to a robot.
